course3/homework-list/linked-list.h

```diff
@@ -0,0 +1,241 @@
+
+#include <cstddef>
+#include <memory>
+
+template <typename TElement>
+class LinkedList
+{
+    struct Node
+    {
+        Node(const TElement& data) : mData(data), mNext()
+        {
+        }
+
+        Node(const TElement& data, std::unique_ptr<Node>&& next) : mData(data), mNext(std::move(next))
+        {
+        }
+
+        Node(TElement&& data, std::unique_ptr<Node>&& next) : mData(std::move(data)), mNext(std::move(next))
+        {
+        }
+
+        const TElement& data() const
+        {
+            return mData;
+        }
+
+        TElement& data()
+        {
+            return mData;
+        }
+
+        Node* next()
+        {
+            return mNext.get();
+        }
+
+        Node* insert_after(const TElement& data)
+        {
+            mNext = std::move(std::make_unique<Node>(data, std::move(mNext)));
+
+            return mNext.get();
+        }
+
+        void eraseNext()
+        {
+            auto& tmp = mNext;
+            mNext = std::move(mNext->mNext);
+        }
+
+        std::unique_ptr<Node>&& remove()
+        {
+            return std::move(mNext);
+        }
+        
+      private:
+        TElement mData;
+        std::unique_ptr<Node> mNext;
+    };
+
+    struct Iterator
+    {
+        using value_type = TElement;
+        using pointer = TElement*;
+        using reference = TElement&;
+        using difference_type = size_t;
+        using iterator_category = std::forward_iterator_tag;
+
+        Iterator() : mPointer(nullptr)
+        {
+        }
+
+        Iterator(const Iterator& other) : mPointer(other.mPointer)
+        {
+        }
+
+        Iterator(const std::unique_ptr<Node>& node) : mPointer(node.get())
+        {
+        }
+
+        bool operator!=(const Iterator& that)
+        {
+            return that.mPointer != mPointer;
+        }
+        
+        bool operator==(const Iterator& that)
+        {
+            return that.mPointer == mPointer;
+        }
+                
+        const TElement& operator*()
+        {
+            return mPointer->data();
+        }
+
+        Iterator& operator++()
+        {
+            mPointer = mPointer->next();
+            return *this;
+        }
+
+        Iterator operator++ (int)
+        {
+            // FIXME: wrong order
```

Comment:
I fixed this but forgot to commit it, clearly I don't have enough unit tests yet.